[SolidMechanics] TetrahedronFEMForceField: adds error message for Poisson's ratio

EulalieCoevoet opened the pull request:
Poisson's ratio should be in [0,0.5[. This PR adds an error message if it's not the case. 
A value outside [0,0.5[ just leads to an odd behavior. 

**EDIT**: a warning message (not error)



______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

fredroy committed:
[SofaNG] Setup Engine (#2812)

* create (emtpy) engine and its subs

* move files

* update includes and API

* add dependency for engine.geometry

* update namespaces

* add compat layer

* update component change list

* fix compilation

* remove forgotten vc characters

* re-dispatch into analyze, generate, select and transform

EulalieCoevoet committed:
[SolidMechanics] TetrahedronFEMForceField: adds an error message if Poisson's ratio is not in [0,0.5[

fredroy commented on Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedronFEMForceField.inl:
```diff
@@ -1349,6 +1349,12 @@ void TetrahedronFEMForceField<DataTypes>::init()
+    const Real& poissonRatio = _poissonRatio.getValue();
+    if (poissonRatio<0 || poissonRatio>=0.5)
+    {
+        msg_error() << "FEM Poisson Ratio in Hooke's law should be in [0,0.5[.";
```
I guess something has to be done if the user encounters this case.

- Either setting the component to invalid (with componentstate) 
- or forcefully clamp the value between [0,0.5[ (but you are purposely modifying its simulation so maybe not the best solution)

alxbilger replied:
In theory, negative Poisson's ratios are possible. But I don't know if it's possible in this component

Refs:

- https://en.wikipedia.org/wiki/Auxetics
- https://en.wikipedia.org/wiki/Poisson%27s_ratio

fredroy replied:
I will let other people (more versed in mechanics than me ) to answer 😏
@hugtalbot maybe ?

EulalieCoevoet replied:
FYI I've just copied the data's info. I don't think we should use a negative ratio with Hooke's law. https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.661.2128&rep=rep1&type=pdf

```cpp
Data<Real> _poissonRatio; ///< FEM Poisson Ratio [0,0.5[
```
```cpp
_poissonRatio(initData(&_poissonRatio,(Real)0.45f,"poissonRatio","FEM Poisson Ratio in Hooke's law [0,0.5["))
```

To be honest my first intention was to set the component to an invalid state but I got discouraged by the 45 methods to identify/update. In the current implementation adding a `return;` after the error message (which would set the state of the component to invalid) will just make the simulation crash, since it is not used at all elsewhere. 
I don't think that clamping the value is a good idea either, same as for setting the default value.

EulalieCoevoet replied:
FYI we need at least a message because of this -> https://github.com/SofaDefrost/SoftRobots/discussions/130 (nan values if you set the ratio to 0.5).

EulalieCoevoet committed:
[SolidMechanics] warning instead of error message for a Poisson's ratio out of bound

alxbilger commented on Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedronFEMForceField.inl:
```diff
@@ -1349,6 +1349,12 @@ void TetrahedronFEMForceField<DataTypes>::init()
+    if (poissonRatio<0 || poissonRatio>=0.5)
+    {
+        msg_warning() << "FEM Poisson Ratio in Hooke's law should be in [0,0.5[.";
+    }
```
```suggestion
    if (poissonRatio < 0 || poissonRatio >= 0.5)
    {
        msg_warning() << "FEM Poisson's Ratio in Hooke's law should be in [0,0.5[.";
    }
```

damienmarchal replied:
The two choice should be either:
- a warning + clamping the value to the closest boundary
- an error and setting the component to invalid

But please don't mix warning with invalid component state.

alxbilger commented:
_Suggestion_
There is also the option to limit the range of the Poisson's ratio in the UI:

![image](https://user-images.githubusercontent.com/10572752/166240168-0be5f1c7-f857-42fe-94be-7cc11be83b29.png)

You need to add `_poissonRatio.setWidget("poissonRatio");` in the constructor of the component.
TetrahedralCorotationalFEMForceField is the only component using this widget.

alxbilger committed:
[FEM.Elastic] Fix typo on Poisson's ratio (#2911)

an delegate constructor

maddin200 committed:
[applications] remove shadow variable (#2910)

fredroy committed:
[S.C.Engine] Remove previous engine files and fix some cmake.in (#2909)

EulalieCoevoet committed:
Update Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedronFEMForceField.inl

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

EulalieCoevoet committed:
clamping

EulalieCoevoet commented:
poke @hugtalbot

fredroy committed:
[Sofa.GL.Component] OglModel: Uniformize floating point type (#2856)

* fix tangents/bitangets type set for opengl

* clean and uniformize floating point type

alxbilger commented:
[ci-build][with-all-tests]

fredroy committed:
[SofaNG] Re-organize unit tests (#2873)

Based on #2813 

and sofang-ize paths/plugins/etc.
and creation of some *Testing framework (solidmechanics, topology, mapping...)

and a loooots of clean as well. (not enough)

Some tests were disabled, they are back again

maddin200 committed:
[applications] remove shadow variables (#2921)

* [applciations] remove shadow variables

* Remove commented code

Co-authored-by: Alex Bilger <alexbilger0@gmail.com>

fredroy committed:
[Sofa.Component] Fix compilation when FLOATING_POINT is set to float (#2907)

* fix compilation with float as floating_point type

* change pragma to warning

* Apply suggestions from code review (typos and comment removal)

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

* fix preprocessor test on sofa_float

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

alxbilger committed:
[Constraint.Lagrangian.Solver] Convert double to SReal (#2922)

guparan committed:
Update README.md (#2905)

fredroy committed:
[SofaNG] Re-organize Framework (#2876)

* move and clean framework modules

* continue with simulation

* aliases and deprecated modules

* clean ide folder, add compat

* sofa.simulation modules compile

* fix various stuff for compat with sofasimulation maze

* clean, compile, sofacreator removal from framework, etc

* fix units tests

* move folders, apply new modules name

* revert sofabasevisual removal (need s.c.userinterface first)

* fix compilation

* refresh pr

* new organization, and cleanings

* add back obj files for testing (gitignore ignores obj...)

* add meta module sofa.simulation

Co-authored-by: Frédérick ROY <fred.roy.mr@gmail.com>

bakpaul committed:
[SofaCuda] Forward declaration in wrong namespace (#2923)

* [SofaCuda] Forward declaration in wrong namespace

* Added changes for double

Co-authored-by: p.baksic <p.baksic@unistra.fr>

alxbilger commented:
Could you solve the conflicts?

EulalieCoevoet committed:
fix conflicts
